### ISSUE TITLE
Optimise the remote reindex for higher throughput

### ISSUE THREAD

**Issue by ankitkala** [opensearch-project/OpenSearch#3765]
**Is your feature request related to a problem? Please describe.**
Current implementation of Remote Reindex is very slow. This issue is to track the improvements we want to do to improve the re-indexing speed.

**Describe the solution you'd like**
We don't have a proposal ready yet. Will update this issue when the proposal ready.

**Comment by lewijacn** [comment]
@ankitkala I have been poking around this area as well recently, would love to hear any thoughts you might have on potential improvements we could make

**Comment by ankitkala** [comment]
On #2, I think there were 2 issues that i observed: 
1. Reindex process uses apache http async client for operation. Due to some bug in handling callback, reindex process was ending up getting executed on the http callback thread instead of the original thread for reindexing.
2. Regarding the halts in remote reindex process, even though the remote cluster returns the scroll response, the reindex process didn't get the callback(need to see if its related for point mentioned above)

**Comment by lewijacn** [comment]
Much thanks for listing this out, it's very helpful 

I definitely agree on point 1, this was one of my own points I've been collecting. It would be a big win to have some form of parallelization on remote reindexing. I don't have a good understanding yet on our limitations with our current implementation for reindexing and what would need to be added/changed to allow for this, but this may be a good place to do further digging.

For point 2, this is something I haven't seen yet in issues or my own testing (likely from testing with smaller datasets). If we could capture a repeatable instance of this, it may make a good smaller hit issue like you mentioned.

A couple of other thoughts that have come to mind that I have yet to dive into:

a. Is there any further compression we can do to reduce the data load we are transferring
b. In some respects, it may be more beneficial for users currently to want to create a snapshot and restore it on a remote cluster and then perform their reindexing locally to take advantage of slicing as well as minimize this large transfer time. Somewhere in this union of approaches I think there may be some good wins to explore
c. By default we are slicing on the `_id` field, which may not always provide similar slicing loads. Potentially adding some more customization to our slicing could be beneficial (Please correct me if this already exists)

**Comment by ankitkala** [comment]
> I definitely agree on point 1, this was one of my own points I've been collecting. It would be a big win to have some form of parallelization on remote reindexing. I don't have a good understanding yet on our limitations with our current implementation for reindexing and what would need to be added/changed to allow for this, but this may be a good place to do further digging.

Yeah. Current implementation simply fetches all the data and does a local bulk which will take care of splitting the docs to multiple shards. If we want to explore parallelisation, we'll need each node to only fetch the data to be indexed on that node. 
Just thinking out loud here, maybe all nodes can start calling the same scroll query for each shard on the node to achieve the parallelisation. But this might not guarantee that data is evenly distributed across all shards though.

> Is there any further compression we can do to reduce the data load we are transferring

Haven't explored this. My guess is the nw bandwidth might not be the bottleneck here. Compression would definitely increase the CPU overhead on the local domain's node which can become an issue. 

> For point 2, this is something I haven't seen yet in issues or my own testing (likely from testing with smaller datasets). If we could capture a repeatable instance of this, it may make a good smaller hit issue like you mentioned.

Yeah. I also haven't played around with remote reindex recently. Let me explore whether this problem still exists. 

> it may be more beneficial for users currently to want to create a snapshot and restore it on a remote cluster 

One benefit of remote reindex is that we can reindex across different OS/ES versions(e.g. ES 1.5 to ES 7.10). With snapshot we can fasten this process but we lose on the versatility of reindexing across versions. Also, since the reindex process happens over rest(local cluster calls remote cluster over http just like any other client would), building the support for fetching segments(snapshot) would be  tricky.

> default we are slicing on the _id field

My understanding is that remote reindex doesn't support slicing(only local reindex).

**Comment by rohanmeh** [comment]
Hi, have there been any improvements made to remote reindexing speed? I am seeing roughly 8 mb/s reindexing speed, which is in the range of the benchmarking mentioned by @ankitkala above. If I do multiple concurrent reindexing tasks I haven't been able to increase the throughput any further, was hoping that may work.
